src/foam/nanos/jetty/WhitelistedForwardedRequestCustomizer.java

```diff
@@ -0,0 +1,210 @@
+/**
+ * @license
+ * Copyright 2019 The FOAM Authors. All Rights Reserved.
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package foam.nanos.jetty;
+
+import org.eclipse.jetty.http.*;
+import org.eclipse.jetty.server.*;
+import org.eclipse.jetty.util.HostPort;
+
+import java.net.InetSocketAddress;
+import java.util.Set;
+import java.util.regex.Pattern;
+
+import foam.core.X;
+import foam.nanos.logger.Logger;
+
+import javax.ws.rs.NotAuthorizedException;
+
+/**
+ * A whitelist-featured customizer which looks at an HTTP request for headers which indicate it
+ * has been forwarded by one or more proxies. Currently, this customizer can only the handle X-Forwarded-For
+ *
+ * If a whitelist is set up and the X-Forwarded-For header exists:
+ *    the customizer will then check most recent IP address of the request against the whitelist and if it does pass,
+ *    the remoteAddress of the request will be set to the address in X-Forwarded-For.
+ * If a whitelist is not set and the X-Forwarded0For header exists:
+ *    the customizer will log the IP address of the request and fail the request
+ * If NO whitelist is set up:
+ *    then all requests with the X-Forwarded-For header will have their remoteAddress set to the address in X-Forwarded-For
+ * Requests WITHOUT the X-Forwarded-For header will be unaffected by this customizer
+ *
+ * For more info on X-Forwarded-For:
+ * @see <a href="http://en.wikipedia.org/wiki/X-Forwarded-For">Wikipedia: X-Forwarded-For</a>
+ *
+ * For implementation of ForwardedRequestCustomizer as per Jetty version 9.4.x:
+ * @see <a href="https://github.com/eclipse/jetty.project/blob/jetty-9.4.x/jetty-server/src/main/java/org/eclipse/jetty/server/ForwardedRequestCustomizer.java">Jetty: ForwardedRequestCustomizer</a>
+ */
+public class WhitelistedForwardedRequestCustomizer extends ForwardedRequestCustomizer
+{
+  private Logger logger;
+  private String _forwardedForHeader = HttpHeader.X_FORWARDED_FOR.toString();
+  private Set<String> forwardedForProxyWhitelist;
+  private X x;
+
+  // these are patterns being declared here to handle proper normalization of IPv6 addresses
+  private static final Pattern IPV6_STD_PATTERN = Pattern.compile("^(?:[0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$");
```

Comment:
@kgrgreer @adamvy do we have a Thread Local Pattern?

Reply:
This is something I can do.  We'll merge now and add a Thread safe pattern in another PR.